java.lang.AssertionError: Element: LatexKeyvalPairImpl(KEYVAL_PAIR); next: DummyHolder; next.prev: null Top:LatexContentImpl(CONTENT)

### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
IntelliJ IDEA 2021.1.1 (build IU-211.7142.45)

### Operating System 
Mac OS X 11.4 (x86_64)

### TeXiFy IDEA version
0.7.6

### Description



### Stacktrace
```
java.lang.AssertionError: Element: LatexKeyvalPairImpl(KEYVAL_PAIR); next: DummyHolder; next.prev: null Top:LatexContentImpl(CONTENT)
	at com.intellij.util.WalkingState.next(WalkingState.java:87)
	at com.intellij.util.WalkingState.walkChildren(WalkingState.java:58)
	at com.intellij.util.WalkingState.elementStarted(WalkingState.java:49)
	at com.intellij.psi.PsiWalkingState.elementStarted(PsiWalkingState.java:76)
	at com.intellij.psi.PsiRecursiveElementWalkingVisitor.visitElement(PsiRecursiveElementWalkingVisitor.java:48)
	at com.intellij.psi.util.PsiTreeUtil$4.visitElement(PsiTreeUtil.java:938)
	at com.intellij.psi.PsiElementVisitor.visitFile(PsiElementVisitor.java:35)
	at com.intellij.psi.PsiRecursiveElementWalkingVisitor.visitFile(PsiRecursiveElementWalkingVisitor.java:70)
	at com.intellij.extapi.psi.PsiFileBase.accept(PsiFileBase.java:60)
	at com.intellij.psi.util.PsiTreeUtil.processElements(PsiTreeUtil.java:934)
	at com.intellij.psi.util.PsiTreeUtil.findChildrenOfAnyType(PsiTreeUtil.java:328)
	at com.intellij.psi.util.PsiTreeUtil.findChildrenOfType(PsiTreeUtil.java:291)
	at nl.hannahsten.texifyidea.util.PsiKt.childrenOfType(Psi.kt:29)
	at nl.hannahsten.texifyidea.folding.LatexMathSymbolFoldingBuilder.buildFoldRegions(LatexMathSymbolFoldingBuilder.kt:21)
	at com.intellij.lang.folding.LanguageFolding.buildFoldingDescriptorsNoPlaceholderCaching(LanguageFolding.java:92)
	at com.intellij.lang.folding.CompositeFoldingBuilder.buildFoldRegions(CompositeFoldingBuilder.java:36)
	at com.intellij.lang.folding.LanguageFolding.buildFoldingDescriptorsNoPlaceholderCaching(LanguageFolding.java:92)
	at com.intellij.lang.folding.LanguageFolding.buildFoldingDescriptors(LanguageFolding.java:75)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.getFoldingsFor(FoldingUpdate.java:267)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.getFoldingsFor(FoldingUpdate.java:233)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.getUpdateResult(FoldingUpdate.java:87)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.lambda$updateFoldRegions$0(FoldingUpdate.java:76)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:54)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:111)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:43)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.updateFoldRegions(FoldingUpdate.java:73)
	at com.intellij.codeInsight.folding.impl.CodeFoldingManagerImpl.updateFoldRegions(CodeFoldingManagerImpl.java:222)
	at com.intellij.codeInsight.folding.impl.CodeFoldingManagerImpl.updateFoldRegionsAsync(CodeFoldingManagerImpl.java:208)
	at com.intellij.codeInsight.daemon.impl.CodeFoldingPass.doCollectInformation(CodeFoldingPass.java:42)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:56)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$1(PassExecutorService.java:400)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1096)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:393)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:688)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:634)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:64)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:392)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:368)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:167)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:178)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:366)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:188)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

--- a/src/nl/hannahsten/texifyidea/psi/LatexPsiHelper.kt
+++ b/src/nl/hannahsten/texifyidea/psi/LatexPsiHelper.kt
@@ -2,6 +2,7 @@ package nl.hannahsten.texifyidea.psi
 
 import com.intellij.openapi.project.Project
 import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
 import com.intellij.psi.PsiFileFactory
 import com.intellij.psi.impl.source.tree.LeafPsiElement
 import nl.hannahsten.texifyidea.LatexLanguage
@@ -10,6 +11,11 @@ import nl.hannahsten.texifyidea.util.childrenOfType
 import nl.hannahsten.texifyidea.util.findFirstChild
 import nl.hannahsten.texifyidea.util.firstChildOfType
 
+/**
+ * As the IntelliJ SDK docs say, to replace or insert text it is easiest to create a dummy file,
+ * fill it with the desired text, extract the psi elements and put them in the actual psi tree.
+ * Don't use document.insertString because that will lead to various things (like inspections) working with an obsolete psi tree.
+ */
 class LatexPsiHelper(private val project: Project) {
 
     private fun createEnvironmentContent(): LatexEnvironmentContent {
@@ -43,7 +49,10 @@ class LatexPsiHelper(private val project: Project) {
         return optionalParam.keyvalPairList[0]
     }
 
-    fun createFromText(text: String): PsiElement =
+    /**
+     * Create a PsiFile containing the given text.
+     */
+    fun createFromText(text: String): PsiFile =
         PsiFileFactory.getInstance(project).createFileFromText("DUMMY.tex", LatexLanguage.INSTANCE, text, false, true)
 
     /**
@@ -130,7 +139,8 @@ class LatexPsiHelper(private val project: Project) {
                 existing
             }
             else {
-                optionalParam.addBefore(createFromText(","), closeBracket)
+                val comma = createFromText(",").firstChildOfType(LatexNormalText::class)?.firstChild ?: return pair
+                optionalParam.addBefore(comma, closeBracket)
                 optionalParam.addBefore(pair, closeBracket)
                 closeBracket.prevSibling as LatexKeyvalPair
             }

--- a/test/nl/hannahsten/texifyidea/inspections/latex/codestyle/LatexMissingLabelInspectionTest.kt
+++ b/test/nl/hannahsten/texifyidea/inspections/latex/codestyle/LatexMissingLabelInspectionTest.kt
@@ -3,6 +3,8 @@ package nl.hannahsten.texifyidea.inspections.latex.codestyle
 import nl.hannahsten.texifyidea.file.LatexFileType
 import nl.hannahsten.texifyidea.inspections.TexifyInspectionTestBase
 import nl.hannahsten.texifyidea.lang.CommandManager
+import nl.hannahsten.texifyidea.psi.LatexKeyvalPair
+import nl.hannahsten.texifyidea.util.childrenOfType
 
 class LatexMissingLabelInspectionTest : TexifyInspectionTestBase(LatexMissingLabelInspection()) {
 
@@ -160,20 +162,24 @@ class LatexMissingLabelInspectionTest : TexifyInspectionTestBase(LatexMissingLab
         """.trimIndent()
     )
 
-    fun `test quick fix in listings with other parameters`() = testQuickFix(
-        before = """
-        \begin{document}
-            \begin{lstlisting}[someoption,otheroption={with value}]
-            \end{lstlisting}
-        \end{document}
-        """.trimIndent(),
-        after = """
-        \begin{document}
-            \begin{lstlisting}[someoption,otheroption={with value},label={lst:lstlisting}]
-            \end{lstlisting}
-        \end{document}
-        """.trimIndent()
-    )
+    fun `test quick fix in listings with other parameters`() {
+        testQuickFix(
+            before = """
+                \begin{document}
+                    \begin{lstlisting}[someoption,otheroption={with value}]
+                    \end{lstlisting}
+                \end{document}
+            """.trimIndent(),
+            after = """
+                \begin{document}
+                    \begin{lstlisting}[someoption,otheroption={with value},label={lst:lstlisting}]
+                    \end{lstlisting}
+                \end{document}
+            """.trimIndent()
+        )
+        // Sometimes, errors in psi structure only show when initiating a WalkingState
+        myFixture.file.children.first().childrenOfType(LatexKeyvalPair::class)
+    }
 
     fun `test fix all missing label problems in this file`() = testQuickFixAll(
             before = """

